File: app/src/main/java/com/arcao/geocaching4locus/settings/fragment/filter/DifficultyFilterPreferenceFragment.java
package com.arcao.geocaching4locus.settings.fragment.filter;

import android.content.SharedPreferences;
import android.os.Bundle;

import com.arcao.geocaching4locus.R;
import com.arcao.geocaching4locus.base.fragment.AbstractPreferenceFragment;
import com.arcao.geocaching4locus.settings.widget.ListPreference;

public class DifficultyFilterPreferenceFragment extends AbstractPreferenceFragment {
    @Override
    public void onCreate(Bundle paramBundle) {
        super.onCreate(paramBundle);

        // Load the preferences from an XML resource
        addPreferencesFromResource(R.xml.preference_category_filter_difficulty);
    }

    @Override
    protected void preparePreference() {
        final ListPreference difficultyMinPreference = findPreference(FILTER_DIFFICULTY_MIN, ListPreference.class);
        final ListPreference difficultyMaxPreference = findPreference(FILTER_DIFFICULTY_MAX, ListPreference.class);

        difficultyMinPreference.setSummary(prepareRatingSummary(difficultyMinPreference.getValue()));
        difficultyMinPreference.setOnPreferenceChangeListener((preference, newValue) -> {
            float min = Float.parseFloat((String) newValue);
            float max = Float.parseFloat(difficultyMaxPreference.getValue());

            if (min > max) {
                difficultyMaxPreference.setValue((String) newValue);
                difficultyMaxPreference.setSummary(prepareRatingSummary((CharSequence) newValue));
            }
            return true;
        });

        difficultyMaxPreference.setSummary(prepareRatingSummary(difficultyMaxPreference.getValue()));
        difficultyMaxPreference.setOnPreferenceChangeListener((preference, newValue) -> {
            float min = Float.parseFloat(difficultyMinPreference.getValue());
            float max = Float.parseFloat((String) newValue);

            if (min > max) {
                difficultyMinPreference.setValue((String) newValue);
                difficultyMinPreference.setSummary(prepareRatingSummary((CharSequence) newValue));
            }
            return true;
        });
    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        super.onSharedPreferenceChanged(sharedPreferences, key);

        if (key == null)
            return;

        switch (key) {
            case FILTER_DIFFICULTY_MIN:
            case FILTER_DIFFICULTY_MAX:
                final ListPreference difficultyTerrainPreference = findPreference(key, ListPreference.class);
                difficultyTerrainPreference.setSummary(prepareRatingSummary(difficultyTerrainPreference.getEntry()));
                break;
        }
    }

    CharSequence prepareRatingSummary(CharSequence value) {
        return preparePreferenceSummary(value, 0);
    }
}

File: app/src/main/java/com/arcao/geocaching4locus/settings/fragment/filter/DifficultyFilterPreferenceFragment.kt
package com.arcao.geocaching4locus.settings.fragment.filter

import android.content.SharedPreferences
import androidx.preference.ListPreference
import com.arcao.geocaching4locus.R
import com.arcao.geocaching4locus.base.constants.PrefConstants.FILTER_DIFFICULTY_MAX
import com.arcao.geocaching4locus.base.constants.PrefConstants.FILTER_DIFFICULTY_MIN
import com.arcao.geocaching4locus.base.fragment.AbstractPreferenceFragment

class DifficultyFilterPreferenceFragment : AbstractPreferenceFragment() {
    override val preferenceResource: Int
        get() = R.xml.preference_category_filter_difficulty

    override fun preparePreference() {
        super.preparePreference()

        val difficultyMinPreference = preference<ListPreference>(FILTER_DIFFICULTY_MIN)
        val difficultyMaxPreference = preference<ListPreference>(FILTER_DIFFICULTY_MAX)

        difficultyMinPreference.summary = prepareRatingSummary(difficultyMinPreference.value)
        difficultyMinPreference.setOnPreferenceChangeListener { _, newValue ->
            val min = (newValue as String).toFloat()
            val max = difficultyMaxPreference.value.toFloat()

            if (min > max) {
                difficultyMaxPreference.value = newValue
                difficultyMaxPreference.summary = prepareRatingSummary(newValue as CharSequence)
            }
            true
        }

        difficultyMaxPreference.summary = prepareRatingSummary(difficultyMaxPreference.value)
        difficultyMaxPreference.setOnPreferenceChangeListener { _, newValue ->
            val min = difficultyMinPreference.value.toFloat()
            val max = (newValue as String).toFloat()

            if (min > max) {
                difficultyMinPreference.value = newValue
                difficultyMinPreference.summary = prepareRatingSummary(newValue as CharSequence)
            }
            true
        }
    }

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String) {
        super.onSharedPreferenceChanged(sharedPreferences, key)

        when (key) {
            FILTER_DIFFICULTY_MIN,
            FILTER_DIFFICULTY_MAX -> {
                preference<ListPreference>(key).apply {
                    summary = prepareRatingSummary(entry ?: return)
                }
            }
        }
    }

    private fun prepareRatingSummary(value: CharSequence): CharSequence {
        return preparePreferenceSummary(value, 0)
    }
}
